candle-metal-kernels/src/metal/command_buffer_pool.rs

```diff
@@ -0,0 +1,255 @@
+    ) -> Result<Arc<CommandBufferEntry>, MetalKernelError> {
+        let semaphore = Arc::new(CommandSemaphore::new());
+        let cb = create_command_buffer(command_queue, Arc::clone(&semaphore))?;
+
+        Ok(Arc::new(CommandBufferEntry {
+            state: Mutex::new(EntryState {
+                current: cb,
+                in_flight: Vec::new(),
+            }),
+            compute_count: AtomicUsize::new(0),
+            semaphore,
+        }))
+    }
+
+    pub fn acquire_compute_encoder(
+        &self,
+    ) -> Result<(bool, ComputeCommandEncoder), MetalKernelError> {
+        let entry = self.select_entry()?;
+        let (flushed, encoder) = self.finalize_entry(entry, |cb| cb.compute_command_encoder())?;
+        Ok((flushed, encoder))
+    }
+
+    pub fn acquire_blit_encoder(&self) -> Result<(bool, BlitCommandEncoder), MetalKernelError> {
+        let entry = self.select_entry()?;
+        let (flushed, encoder) = self.finalize_entry(entry, |cb| cb.blit_command_encoder())?;
+        Ok((flushed, encoder))
+    }
+
+    /// Selects an entry from the pool using a two-phase strategy:
+    /// 1. Try non-blocking: find any available buffer without waiting
+    /// 2. Fallback: select the least-loaded buffer and wait for availability
+    fn select_entry(&self) -> Result<Arc<CommandBufferEntry>, MetalKernelError> {
+        // Phase 1: Try to find an available buffer without blocking
+        for entry in &self.pool {
+            if let Ok(mut status) = entry.semaphore.status.try_lock() {
+                if matches!(*status, CommandStatus::Available) {
+                    *status = CommandStatus::Encoding;
+                    return Ok(Arc::clone(entry));
+                }
+            }
+        }
+
+        // Phase 2: Select the buffer with the least work and wait for it
+        let entry = self
+            .pool
+            .iter()
+            .min_by_key(|e| e.compute_count.load(Ordering::Acquire))
+            .ok_or(MetalKernelError::FailedToCreateResource(
+                "Command buffer pool is empty".to_string(),
+            ))?;
+
+        let entry = Arc::clone(entry);
+        {
+            let mut guard = entry
+                .semaphore
+                .wait_until(|s| matches!(s, CommandStatus::Available));
+            *guard = CommandStatus::Encoding;
+        }
+
+        Ok(entry)
+    }
+
+    /// Creates an encoder from the selected entry, recycling the buffer if needed.
+    /// When recycling, the old committed buffer is moved to `in_flight` so we can later wait on it.
+    fn finalize_entry<F, E>(
+        &self,
+        entry: Arc<CommandBufferEntry>,
+        create_encoder: F,
+    ) -> Result<(bool, E), MetalKernelError>
+    where
+        F: FnOnce(&mut CommandBuffer) -> E,
+    {
+        let mut state = entry.state.lock()?;
+
+        let count = entry.compute_count.fetch_add(1, Ordering::Relaxed);
+        let flushed = count >= self.compute_per_buffer;
+
+        if flushed {
```

Comment:
Nit. To me `flushed` in past tense implies "we have flushed" or "has been flushed", so for if-statements like this I prefer "should we flush?"
```suggestion
        let flush = count >= self.compute_per_buffer;

        if flush {
```